### PR TITLE
Support BLS Aggregate Signatures in FVM (FIP-0079)

### PR DESCRIPTION
**Pull request by DrPeterVanNostrand**
FIP Discussion: #840

### CODE REVIEW

File: FIPS/fip-00xx.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+) -> bool;
+```
+
+### Gas Pricing
+
+The gas pricing for `verify_bls_aggregate` is dependent upon the number of signatures aggregated and the total size (in bytes) of the signed plaintexts. The number of signatures aggregated determines the number of elliptic curve pairing operations performed by the signature verifier, and the total size of plaintexts determines the amount of hashing that a verifier must perform when mapping each plaintext to a curve point.
+
+The gas cost of BLS aggregate signature verification is computed via:
+
+**TODO:** add real gas costs.
```
Comment (anorth) [inline FIPS/fip-00xx.md:79]: There are real-looking numbers below. Please confirm whether this TODO is outstanding (in which case I suggest removing the numbers), or done (remove the TODO).

**Review by anorth — APPROVED** [review]

File: FIPS/fip-00xx.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    miller_loop.final_exponentiation() == Gt::one()
+}
+```
+
+FVM uses the Rust crates [`bls-signatures`](https://github.com/filecoin-project/bls-signatures) for making and verifying BLS signatures, and [`blstrs`](https://github.com/filecoin-project/blstrs) for functionality specific to elliptic curve BLS12-381, which FVM uses for BLS signatures.
+
+## TODO
+<!--A section that lists any unresolved issues or tasks that are part of the FIP proposal. Examples of these include performing benchmarking to know gas fees, validate claims made in the FIP once the final implementation is ready, etc. A FIP can only move to a “Last Call” status once all these items have been resolved.-->
+
+None.
```
Comment (jsoares) [inline FIPS/fip-00xx.md:182]: Inconsistent with the 2 todo items recorded in other sections. Please clarify what is to be done and what are just leftover/stale comments.

**Review by jsoares — APPROVED** [review]
Looks almost good to merge, note few minor comments. 

@anorth I'm not seeing any FIP-0079 in flight. Is that our next number?

File: FIPS/fip-00xx.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+This FIP proposes the following changes to FVM's syscalls:
+
+- Addition of a syscall `verify_bls_aggregate` for verifying a BLS aggregate signature; because the verification algorithm for non-aggregate BLS signatures is identical to that for aggregations of one signature, `verify_bls_aggregate` supports both aggregate and non-aggregate verification. The syscall's API is:
+
+```rust
+fn verify_bls_aggregate(
+    aggregate_signature: &[u8; 96],
+    pub_keys: &[[u8; 48]],
+    plaintexts: &[&[u8]],
+) -> bool;
```
Comment (Stebalien) [inline FIPS/fip-00xx.md:46]: Let's document the exact syscall API (i.e., the low-level FFI signature).

File: FIPS/fip-00xx.md (hunk 1 above)
Comment (Stebalien) [reply FIPS/fip-00xx.md:79]: Those numbers were estimated from existing gas-charges without re-benchmarking.

Comment (Stebalien) [reply FIPS/fip-00xx.md:79]: (and that comment should probably go in the FIP)

File: FIPS/fip-00xx.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+## Security Considerations
+<!--All FIPs must contain a section that discusses the security implications/considerations relevant to the proposed change. Include information that might be important for security discussions, surfaces risks and can be used throughout the life cycle of the proposal. E.g. include security-relevant design decisions, concerns, important discussions, implementation-specific guidance and pitfalls, an outline of threats and risks and how they are being addressed. FIP submissions missing the "Security Considerations" section will be rejected. A FIP cannot proceed to status "Final" without a Security Considerations discussion deemed sufficient by the reviewers.-->
+
+As FVM currently supports non-aggregate BLS signatures, this FIP does not affect FVM security.
+
+One security consideration in using BLS aggregate signatures is that plaintexts, signed by a single aggregate signature, are required to be unique (i.e. multiple BLS private keys cannot sign the same plaintext, then proceed to aggregate those signatures).
```
Comment (Stebalien) [inline FIPS/fip-00xx.md:117]: Hm. This is actually a good point. Can we _validate_ that the plaintexts _are_ unique when validating the signature? Or is that even necessary?

File: FIPS/fip-00xx.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+## Incentive Considerations
+<!--All FIPs must contain a section that discusses the incentive implications/considerations relative to the proposed change. Include information that might be important for incentive discussion. A discussion on how the proposed change will incentivize reliable and useful storage is required. FIP submissions missing the "Incentive Considerations" section will be rejected. An FIP cannot proceed to status "Final" without a Incentive Considerations discussion deemed sufficient by the reviewers.-->
+
+**TODO:** The addition of aggregate signatures to FVM incentivizes the use of aggregate over non-aggregate signatures, as aggregate signature validation is less computationally expensive than verifying multiple signatures individually (and thus has a lower gas cost).
```
[the remark this answers is not in the document]
Comment (Stebalien) [reply FIPS/fip-00xx.md:122]: Honestly, I don't think this FIP has any impact on incentives.

File: FIPS/fip-00xx.md (hunk 4 above)
Comment (jsoares) [reply FIPS/fip-00xx.md:117]: We seem to be doing that already, no?

https://github.com/DrPeterVanNostrand/FIPs/blob/4cb235e892b2f2017925e1e9947b4905a735fe3f/FIPS/fip-00xx.md?plain=1#L144-L151

Comment (Stebalien) [reply FIPS/fip-00xx.md:117]: Ah, I see. Then I don't think this needs to be mentioned in the security section (or rephrased, right now it sounds like a potential security issue).

**Comment by anorth** [comment]
>  I'm not seeing any FIP-0079 in flight. Is that our next number?

Yes I think so. @DrPeterVanNostrand please use FIP-0079 and update the README to link to this new draft.

**Comment by cryptonemo** [comment]
I applied most feedback review here: https://github.com/cryptonemo/FIPs/tree/jake/fvm-bls-sigs

(And more specifically here: https://github.com/filecoin-project/FIPs/commit/fbb7d2e99de2a00c3ffcaf921be1779778aec12b)

When @DrPeterVanNostrand returns, he can finalize for merging.

File: FIPS/fip-00xx.md (hunk 4 above)
Comment (DrPeterVanNostrand) [reply FIPS/fip-00xx.md:117]: > Can we validate that the plaintexts are unique when validating the signature? Or is that even necessary?

I don't think adding the check in FVM is necessary. The plaintext uniqueness check is always done by the `bls-signatures` crate as part of signature verification [here](https://github.com/filecoin-project/bls-signatures/blob/08fcca2728b3ff5be720f7874c780bfa79e66ec2/src/signature.rs#L145).

**Pull request merged**